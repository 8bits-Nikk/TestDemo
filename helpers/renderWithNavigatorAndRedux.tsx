import React from 'react';
import 'react-native';
import createSagaMiddleware from 'redux-saga';
import {applyMiddleware, createStore, Store} from 'redux';
import rootReducer from '../src/App.reducers';
import root from '../src/App.sagas';
import {render} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';
import {Provider} from 'react-redux';

export function renderWithNavigatorAndRedux(
  renderComponent: any,
  reduxEvent?: (store: Store) => void,
) {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));
  sagaMiddleware.run(root);

  if (reduxEvent) {
    reduxEvent(store);
  }

  return render(
    <NavigationContainer>
      <Provider store={store}>{renderComponent}</Provider>
    </NavigationContainer>,
  );
}
