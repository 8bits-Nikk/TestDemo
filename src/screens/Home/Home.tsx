import {useNavigation} from '@react-navigation/native';
import React, {useEffect, useState} from 'react';
import {
  Alert,
  FlatList,
  Image,
  SafeAreaView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import {shallowEqual, useDispatch, useSelector} from 'react-redux';
import {UsersActions, UsersSelectors} from '../../service/users';
import CardView from '../../components/CardView';
import {DateFormats, formatDate} from '../../service/date/formatDate';
import {imgBack} from '../../assets/assets';
import asyncStorage from '../../service/async-storage/asyncStorage';

const Home = () => {
  const navigation = useNavigation();
  const users = useSelector(UsersSelectors.getUsers, shallowEqual);
  const [currentUser, setCurrentUser] = useState({userName: '', password: ''});
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(UsersActions.getUsers(1));
    asyncStorage.getLoginStatus().then(user => {
      setCurrentUser(user);
    });
  }, []);

  const onLogout = () => {
    asyncStorage.resetLoginStatus().then(res => {
      if (res) {
        navigation.reset({routes: [{name: 'LoginScreen'}]});
      } else {
        Alert.alert('Error', 'Something went wrong');
      }
    });
  };

  return (
    <SafeAreaView style={style.body}>
      <View style={style.header}>
        <TouchableOpacity onPress={onLogout}>
          <Image source={imgBack} style={{height: 24, width: 24}} />
        </TouchableOpacity>
        <Text style={style.headerTxt}>Users</Text>
      </View>
      <View>
        <Text testID={'welcome-name'}>Welcome {currentUser.userName}</Text>
        <Text>
          Today's Date: {formatDate(new Date(), DateFormats.DateOnly)}
        </Text>
      </View>
      <View style={style.container}>
        <FlatList
          testID={'flatList-user'}
          data={users}
          ListEmptyComponent={() => (
            <View>
              <Text style={style.headerTxt}>No Items</Text>
            </View>
          )}
          renderItem={({item}) => (
            <CardView
              item={item}
              onPress={() => navigation.navigate('PostScreen')}
            />
          )}
        />
      </View>
    </SafeAreaView>
  );
};

const style = StyleSheet.create({
  body: {
    flex: 1,
    backgroundColor: '#eee',
  },
  header: {
    flexDirection: 'row',
    backgroundColor: '#aabbdd',
    padding: 16,
    alignItems: 'center',
  },
  headerTxt: {
    marginStart: 16,
    alignItems: 'center',
    fontSize: 18,
    fontWeight: 'bold',
  },
  container: {
    margin: 16,
  },
});

export default Home;
