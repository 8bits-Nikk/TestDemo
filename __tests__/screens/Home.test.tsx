import React from 'react';
import {renderWithNavigatorAndRedux} from '../../helpers/renderWithNavigatorAndRedux';
import Home from '../../src/screens/Home/Home';
import {act, cleanup, screen} from '@testing-library/react-native';
import {UsersActions} from '../../src/service/users';

describe('HomeScreen Tests', () => {
  afterEach(() => {
    cleanup();
  });

  test('Render correctly', () => {
    const tree = renderWithNavigatorAndRedux(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render with data', () => {
    const data = [
      {
        id: 2280,
        name: 'Dhanadeepa Guneta',
        email: 'guneta_dhanadeepa@bayer-schmitt.net',
        gender: 'female',
        status: 'inactive',
      },
      {
        id: 2275,
        name: 'Mani Naik Esq.',
        email: 'naik_esq_mani@bergnaum-labadie.io',
        gender: 'male',
        status: 'inactive',
      },
      {
        id: 2274,
        name: 'Vaijayanthi Mishra',
        email: 'mishra_vaijayanthi@mayert.co',
        gender: 'female',
        status: 'active',
      },
      {
        id: 2273,
        name: 'Yoginder Nayar',
        email: 'yoginder_nayar@luettgen-mraz.com',
        gender: 'male',
        status: 'active',
      },
      {
        id: 2272,
        name: 'Trilok Chopra',
        email: 'chopra_trilok@lehner-hansen.name',
        gender: 'male',
        status: 'active',
      },
      {
        id: 2270,
        name: 'Balagopal Embranthiri Ret.',
        email: 'embranthiri_balagopal_ret@cormier.net',
        gender: 'female',
        status: 'active',
      },
      {
        id: 2269,
        name: 'Mrs. Shreya Deshpande',
        email: 'deshpande_shreya_mrs@konopelski.com',
        gender: 'male',
        status: 'inactive',
      },
      {
        id: 2268,
        name: 'Uttam Ahluwalia',
        email: 'ahluwalia_uttam@gislason-luettgen.org',
        gender: 'female',
        status: 'active',
      },
      {
        id: 2267,
        name: 'Aatreya Marar',
        email: 'aatreya_marar@sawayn-beahan.info',
        gender: 'female',
        status: 'inactive',
      },
      {
        id: 2266,
        name: 'Chiranjeev Reddy MD',
        email: 'chiranjeev_md_reddy@mante-bogisich.biz',
        gender: 'male',
        status: 'inactive',
      },
    ];
    renderWithNavigatorAndRedux(<Home />, store =>
      store.dispatch(UsersActions.addUsers(data)),
    );

    const list = screen.queryByTestId('flatList-user');
    expect(list?.props.data).toBe(data);
    expect(list?.props.data.length).toBe(10);
  });
});
