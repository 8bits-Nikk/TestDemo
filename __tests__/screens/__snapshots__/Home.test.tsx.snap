// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen Tests Render correctly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "#eee",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#aabbdd",
        "padding": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Users
    </Text>
  </View>
  <View>
    <Text>
      Today's Date: 
      21/09/2022
    </Text>
  </View>
  <View
    style={
      Object {
        "margin": 16,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={[Function]}
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      testID="flatList-user"
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View>
          <Text
            style={
              Object {
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            No Items
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
